### PR TITLE
HPCC-26435 Security Page user details

### PR DESCRIPTION
**Pull request by jeclrsg**
Signed-off-by: Jeremy Clements <jeremy.clements@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jeclrsg** [comment]
@GordonSmith wanted to mention that AvailablePermissions.tsx wasn't used yet, but wanted to include it. I couldn't figure out how to replicate the dojo version's triggering of a "dgrid-datachange" event to alter the checkboxes.

**Comment by GordonSmith** [comment]
@jeclrsg please retarget for candidate-8.4.x (this will be true for all future PRs)

File: esp/src/src-react/components/AvailablePermissions.tsx [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            editor: "checkbox",
+            editorArgs: { value: true },
+            className: "hpccCentered",
+            autoSave: true,
+            canEdit: function (object, value) {
+                if (object.__hpcc_type === "Permission" || object.AccountName === "Administrators") {
+                    return false;
+                }
+                return true;
+            },
```
Comment (GordonSmith) [inline esp/src/src-react/components/AvailablePermissions.tsx:127]: Change to:
```
return object.__hpcc_type !== "Permission" && object.AccountName !== "Administrators";

```

File: esp/src/src-react/components/AvailablePermissions.tsx [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            renderHeaderCell: function (node) {
+                node.innerHTML = nlsHPCC.DenyWrite;
```
Comment (GordonSmith) [inline esp/src/src-react/components/AvailablePermissions.tsx:129]: `innerText` (unless DenyWrite has html in it?) - better habit from a security point of view.

File: esp/src/src-react/components/AvailablePermissions.tsx [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            width: 54,
+            editor: "checkbox",
+            editorArgs: { value: true },
+            className: "hpccCentered",
+            autoSave: true,
+            canEdit: function (object, value) {
+                if (object.__hpcc_type === "Permission" || object.AccountName === "Administrators") {
+                    return false;
+                }
+                return true;
```
Comment (GordonSmith) [inline esp/src/src-react/components/AvailablePermissions.tsx:142]: Change to
```
return object.__hpcc_type !== "Permission" && object.AccountName !== "Administrators";
```

"Permission" and "Administrators" should be made module const declarations.

File: esp/src/src-react/components/AvailablePermissions.tsx [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            },
+            renderHeaderCell: function (node) {
+                node.innerHTML = nlsHPCC.DenyFull;
```
Comment (GordonSmith) [inline esp/src/src-react/components/AvailablePermissions.tsx:145]: `innerText` (unless DenyFull has html in it?) - better habit from a security point of view.

File: esp/src/src-react/components/AvailablePermissions.tsx [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+            }
+        }, CheckBox)
+    });
+
+    React.useCallback(() => {
+        // gridStore.on("dgrid-datachange", (evt) => {
+        //     console.log(evt);
+        // });
+    }, []);
```
Comment (GordonSmith) [inline esp/src/src-react/components/AvailablePermissions.tsx:154]: Remove commented code (unless there is an issue here which you can't work around?)

File: esp/src/src-react/components/AvailablePermissions.tsx [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+    const buttons = React.useMemo((): ICommandBarItemProps[] => [
+        {
+            key: "refresh", text: nlsHPCC.Refresh, iconProps: { iconName: "Refresh" },
+            onClick: () => refreshTable()
+        },
+    ], [refreshTable]);
+
+    React.useEffect(() => {
+    }, [grid, gridStore, refreshTable]);
```
Comment (GordonSmith) [inline esp/src/src-react/components/AvailablePermissions.tsx:221]: This does nothing?

File: esp/src/src-react/components/UserGroups.tsx [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+import { CommandBar, ContextualMenuItemType, ICommandBarItemProps } from "@fluentui/react";
+import { useConst } from "@fluentui/react-hooks";
+import { scopedLogger } from "@hpcc-js/util";
+import * as Observable from "dojo/store/Observable";
+import { Memory } from "src/Memory";
+import * as WsAccess from "src/ws_access";
+import nlsHPCC from "src/nlsHPCC";
+import { ShortVerticalDivider } from "./Common";
+import { pushUrl } from "src-react/util/history";
+import { HolyGrail } from "src-react/layouts/HolyGrail";
```
Comment (GordonSmith) [inline esp/src/src-react/components/UserGroups.tsx:11]: `src-react` -> `..`

File: esp/src/src-react/components/UserGroups.tsx [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+import { DojoGrid, selector } from "./DojoGrid";
+import { UserAddGroupForm } from "./forms/UserAddGroup";
+
+const logger = scopedLogger("../components/UserGroups.tsx");
```
Comment (GordonSmith) [inline esp/src/src-react/components/UserGroups.tsx:15]: `..` -> `src-react` :)

File: esp/src/src-react/components/UserGroups.tsx [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    //  Grid ---
+    const gridStore = useConst(new Observable(new Memory("name")));
+    const gridSort = useConst([{ attribute: "name", "descending": false }]);
+    const gridQuery = useConst({});
+    const gridColumns = useConst({
+        check: selector({ width: 27, label: " " }, "checkbox"),
+        name: {
+            label: nlsHPCC.GroupName,
+            formatter: function (_name, idx) {
+                return `<a href="#/security/groups/${_name}">${_name}</a>`;
```
Comment (GordonSmith) [inline esp/src/src-react/components/UserGroups.tsx:43]: `_name` should be run through a sanitizer!!!

File: esp/src/src-react/components/UserGroups.tsx [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    }, [selection]);
+
+    const refreshTable = React.useCallback((clearSelection = false) => {
+        WsAccess.UserEdit({
+            request: { username: username }
+        })
+            .then(({ UserEditResponse }) => {
+                if (UserEditResponse?.Groups) {
+                    const groups = UserEditResponse?.Groups?.Group;
+                    console.log(groups);
```
Comment (GordonSmith) [inline esp/src/src-react/components/UserGroups.tsx:66]: Remove console.log

File: esp/src/src-react/components/forms/Fields.tsx [hunk 11]
```diff
@@ -478,6 +478,43 @@ export const TargetFolderTextField: React.FunctionComponent<TargetFolderTextFiel
...
+                .filter(group => group.name !== "Administrators")
+                .map(group => {
+                    return {
+                        key: group.name,
+                        text: group.name
+                    };
+                });
+            groups.unshift({ key: "", text: "" });
+            setGroups(groups);
+        });
```
Comment (GordonSmith) [inline esp/src/src-react/components/forms/Fields.tsx:511]: Log errors

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/src-react/components/AvailablePermissions.tsx (hunk 6 above)
Comment (jeclrsg) [reply esp/src/src-react/components/AvailablePermissions.tsx]: Removed the file from the PR. It was incomplete & mostly just copied directly over from the dojo component at the moment.

File: esp/src/src-react/components/UserGroups.tsx (hunk 9 above)
Comment (jeclrsg) [reply esp/src/src-react/components/UserGroups.tsx:44]: There might be a lot of these across the codebase. Changed this one instance to use just a simple regex to sanitize, but should maybe find a good / robust library for this? Dompurify?

Comment (GordonSmith) [reply esp/src/src-react/components/UserGroups.tsx:44]: Open a ticket and we can investigate which lib will work for us.

**Review by GordonSmith — APPROVED** [review]

File: esp/src/src-react/components/UserGroups.tsx (hunk 9 above)
Comment (GordonSmith) [reply esp/src/src-react/components/UserGroups.tsx:44]: FYI - all React based items are sanitized by default...

**Comment by richardkchapman** [comment]
@jeclrsg  Needs rebasing for a clean merge

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**